shared/src/main/java/uk/gov/di/authentication/shared/state/StateMachine.java

```diff
@@ -74,148 +82,144 @@ public T transition(T from, A action) {
+                .when(LOGGED_IN)
+                .allow(on(SYSTEM_HAS_SENT_MFA_CODE).then(MFA_SMS_CODE_SENT))
+                .when(MFA_SMS_CODE_SENT)
+                .allow(
+                        on(SYSTEM_HAS_SENT_MFA_CODE).then(MFA_SMS_CODE_SENT),
+                        on(USER_ENTERED_VALID_MFA_CODE).then(MFA_CODE_VERIFIED),
+                        on(USER_ENTERED_INVALID_MFA_CODE).then(MFA_CODE_NOT_VALID))
+                .when(MFA_CODE_NOT_VALID)
+                .allow(
+                        on(USER_ENTERED_VALID_MFA_CODE)
+                                .then(UPDATED_TERMS_AND_CONDITIONS)
+                                .ifCondition(userHasNotAcceptedTermsAndConditionsVersion(configurationService.getTermsAndConditionsVersion())),
+                        on(USER_ENTERED_VALID_MFA_CODE).then(MFA_CODE_VERIFIED).byDefault(),
+                        on(USER_ENTERED_INVALID_MFA_CODE).then(MFA_CODE_NOT_VALID),
+                        on(USER_ENTERED_INVALID_MFA_CODE_TOO_MANY_TIMES)
+                                .then(MFA_CODE_MAX_RETRIES_REACHED))
+                .when(MFA_CODE_MAX_RETRIES_REACHED)
+                .deadEnd()
+                .when(MFA_CODE_VERIFIED)
+                .allow(on(SYSTEM_HAS_ISSUED_AUTHORIZATION_CODE).then(AUTHENTICATED))
+                .when(AUTHENTICATED)
+                .allow(on(SYSTEM_HAS_ISSUED_AUTHORIZATION_CODE).then(AUTHENTICATED))
+                .build();
     }
 
     public static class InvalidStateTransitionException extends RuntimeException {}
 
+    public static <T, A, C> Builder<T, A, C> builder() {
+        return new Builder<>();
+    }
+
+    public static class Builder<T, A, C> {
+        private final Map<T, List<Transition<T, A, C>>> states = new HashMap<>();
+
+        public StateRuleBuilder<T, A, C> when(T state) {
+            return new StateRuleBuilder<>(this, state);
+        }
+
+        protected void addStateRule(T state, List<Transition<T, A, C>> transitions) {
+            this.states.put(state, transitions);
+        }
+
+        public StateMachine<T, A, C> build() {
+            return new StateMachine<>(states);
+        }
+    }
+
+    public static class StateRuleBuilder<T, A, C> {
+        private final Builder<T, A, C> stateMachineBuilder;
+        private final T state;
+
+        @SafeVarargs
+        public final Builder<T, A, C> allow(final Transition.Builder<T, A, C>... transitions) {
+            stateMachineBuilder.addStateRule(
+                    state,
+                    Arrays.asList(transitions).stream()
+                            .map(b -> b.build())
+                            .collect(Collectors.toList()));
+            return stateMachineBuilder;
+        }
+
+        protected StateRuleBuilder(Builder<T, A, C> stateMachineBuilder, T state) {
+            this.stateMachineBuilder = stateMachineBuilder;
+            this.state = state;
+        }
+
+        public Builder<T, A, C> deadEnd() {
```

Comment:
could this be changed to final?

Reply:
`final` is a reserved word, so had to use `finalState()`